src/cudadecoder/cuda-decoder-common.h

```diff
@@ -0,0 +1,468 @@
+// very end of the frame)
+// Because keeping a bit more than max_active_ is expected, we add the tolerance
+// so that we can avoid triggering ApplyMaxActiveAndReduceBeam for just a few
+// tokens above the limit
+// at the end of the frame
+#define KALDI_CUDA_DECODER_MAX_ACTIVE_TOLERANCE 0.2
+
+#define KALDI_CUDA_DECODER_DIV_ROUND_UP(a, b) ((a + b - 1) / b)
+
+#define KALDI_CUDA_DECODER_ASSERT(val, recoverable)                     \
+  {                                                                     \
+    if ((val) != true) {                                                \
+      throw CudaDecoderException("KALDI_CUDA_DECODER_ASSERT", __FILE__, \
+                                 __LINE__, recoverable)                 \
+    }                                                                   \
+  }
+// Macro for checking cuda errors following a cuda launch or api call
+#define KALDI_DECODER_CUDA_CHECK_ERROR()                                  \
+  {                                                                       \
+    cudaError_t e = cudaGetLastError();                                   \
+    if (e != cudaSuccess) {                                               \
+      throw CudaDecoderException(cudaGetErrorName(e), __FILE__, __LINE__, \
+                                 false);                                  \
+    }                                                                     \
+  }
+
+#define KALDI_DECODER_CUDA_API_CHECK_ERROR(e)                             \
+  {                                                                       \
+    if (e != cudaSuccess) {                                               \
+      throw CudaDecoderException(cudaGetErrorName(e), __FILE__, __LINE__, \
+                                 false);                                  \
+    }                                                                     \
+  }
+
+#define KALDI_CUDA_DECODER_1D_KERNEL_LOOP(i, n)                \
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < (n); \
+       i += blockDim.x * gridDim.x)
+
+#define KALDI_CUDA_DECODER_1D_BLOCK_OFFSET_KERNEL_LOOP(offset, th_idx, n) \
+  for (int offset = blockIdx.x * blockDim.x, th_idx = threadIdx.x;        \
+       offset < (n); offset += blockDim.x * gridDim.x)
+
+#define KALDI_CUDA_DECODER_IS_LAST_1D_THREAD() (threadIdx.x == (blockDim.x - 1))
+
+#define KALDI_CUDA_DECODER_BATCH_KERNEL_LOOP(i, n) \
+  for (int i = blockIdx.y; i < (n); i += gridDim.y)
+
+#define KALDI_CUDA_DECODER_DIV_ROUND_UP(a, b) ((a + b - 1) / b)
+
+#define KALDI_CUDA_DECODER_1D_BLOCK 256
+#define KALDI_CUDA_DECODER_LARGEST_1D_BLOCK 1024
+#define KALDI_CUDA_DECODER_ONE_THREAD_BLOCK 1
+
+namespace kaldi {
+namespace cuda_decoder {
+
+// Returning the number of CTAs to launch for (N,M) elements to compute
+// M is usually the batch size
+inline dim3 KaldiCudaDecoderNumBlocks(int N, int M) {
+  dim3 grid;
+  // TODO MAX_NUM_BLOCKS.
```

Comment:
Why would we need a max number of blocks? Not all blocks need to be resident on the GPU at once, anyway.

Reply:
There's a few reasons:
- some kernels aggregate values locally in shared memory first (like the histogram kernel). If the same CTA is in charge of more data (iterating on thread_idx += gridDim.x*blockDim.x), using the shared memory for local aggregates, it will push only once the local aggregates to the global memory. Those global aggregates are atomics so it's a big deal.
- Some data is loaded only once in registers and reused when the CTA loops on thread_idx += gridDim.x*blockDim.x
- Lauching/terminating CTAs has a cost
- More specific to our case is: the amount of work we have to do for each frame depends on that frame (and the previous ones). Which means that each utterances in a batch will have a different number of tokens to process for instance. Because we launch a 2D grid, limiting the number of CTAs allow some load balancing to appear (if we always use the max to decide the x-dimension, all other utterances in the batch will launch CTAs at the end for nothing).

All of this is not important when we generate only a few thousands tokens / frame, but it really speeds up the kernels for challenging models (for this, we usually test on the aspire model evaluated on librispeech/test_other). It gives around 10x on histogram, 2x on expand, etc. 

And for older arch, max # of CTAs / dim[x|y] is ~65000. Even if this is very high for us it's still cleaner to put a clear upper limit